### PR TITLE
Intial skeleton

### PR DESCRIPTION
**Pull request by stevenjack**
### Problem

We want to be able to perform an `Mass Payment` `NVP` request and parse the response.

### Solution

* Add `payload/mass_payment.go` struct for creating the mass payment and adding transactions.
* Serialize the payload into `KEY=VALUE` pairs as per the NVP spec.
* Perform request and parse response into `Response` struct.

### Notes

- **Remember** to tag your merge commit after merging 👍

### CODE REVIEW

**Comment by stevenjack** [comment]
[[GPN 0.2.0] Adds initial skeleton](https://trello.com/c/pMekYhKO/898-gpn-0-2-0-adds-initial-skeleton)

File: client.go [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+import (
+	"bytes"
+	"fmt"
+	"net/http"
+
+	"github.com/vidsy/go-paypalnvp/payload"
+)
+
+const (
+
```
Comment (revett) [inline client.go:15]: Newline

File: client.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+	baseAPIEndpoint = "https://%s.paypal.com/nvp"
+
+	// Sandbox prefix for api signature requests.
+	sandboxAPISignatureRequestPrefix = "api.sandbox"
+
+	// Live prefix for api signature requests.
+	apiSignatureRequestPrefix = "api"
+
+	// Sandsbox environment
+	Sandbox = "sandbox"
```
Comment (revett) [inline client.go:20]: 🏄 Move exported to top

File: client.go [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+	"github.com/vidsy/go-paypalnvp/payload"
+)
+
+const (
+
+	// Base API endpoint for NVP requests.
+	baseAPIEndpoint = "https://%s.paypal.com/nvp"
+
+	// Sandbox prefix for api signature requests.
+	sandboxAPISignatureRequestPrefix = "api.sandbox"
```
Comment (revett) [inline client.go]: Unexported do not need comments.

File: payload/mass_payment.go [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+	"errors"
+	"fmt"
+	"net/url"
+	"reflect"
+	"strconv"
+)
+
+const (
+	// receiverTypeEmail sets receiver type to email address.
+	ReceiverTypeEmail = "EmailAddress"
```
Comment (revett) [inline payload/mass_payment.go:13]: Comments need to be caps - please run go-lint

File: payload/mass_payment.go [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+	// receiverTypePhone sets receiver type to phone number.
+	ReceiverTypePhone = "PhoneNumber"
+
+	// receiverTypeUserID sets receiver type to user id.
+	ReceiverTypeUserID = "UserID"
+)
+
+type (
+	//MassPayment payload for mass payment request.
```
Comment (revett) [inline payload/mass_payment.go]: Space after `//`

File: payload/mass_payment.go [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+)
+
+// NewMassPayment creates a new MassPayment struct with defaults
+func NewMassPayment(user string, password string, signature string, receiverType string) *MassPayment {
+	return &MassPayment{
+		User:         user,
+		Password:     password,
+		Signature:    signature,
+		Method:       "MassPay",
+		CurrencyCode: "GBP",
```
Comment (revett) [inline payload/mass_payment.go]: Hard-coded currency code?

File: payload/mass_payment.go [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+
+// AddItem Adds an item to the mass payment items array.
+func (mp *MassPayment) AddItem(item MassPaymentItem) {
+	mp.Items = append(mp.Items, item)
+}
+
+// Serialize Convert struct into NVP key=value format for the masspayment.
+func (mp MassPayment) Serialize() (string, error) {
+	data := url.Values{}
+	mpType := reflect.TypeOf(mp)
```
Comment (revett) [inline payload/mass_payment.go]: Use proper variable name #verbose

File: payload/mass_payment.go [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+	}
+
+	if len(mp.Items) == 0 {
+		return "", errors.New("Expected at least one mass payment item")
+	}
+
+	return data.Encode(), nil
+}
+
+// Serialize Convert mass payment item into key=value pair and add to existing
```
Comment (revett) [inline payload/mass_payment.go]: 🏄 convert does not need cap

File: payload/mass_payment_test.go [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+	"testing"
+
+	"github.com/vidsy/go-paypalnvp/payload"
+)
+
+func TestMassPayment(t *testing.T) {
+	t.Run(".AddItem", func(t *testing.T) {
+		t.Run("AddsToItemArray", func(t *testing.T) {
+			item := payload.MassPaymentItem{}
+			massPayment := payload.NewMassPayment("user", "password", "signature", payload.ReceiverTypeEmail)
```
Comment (revett) [inline payload/mass_payment_test.go]: 🏄 multi-line

Comment (revett) [reply payload/mass_payment_test.go]: Same for below as well 👍

File: payload/mass_payment_test.go [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+
+			if err == nil {
+				t.Fatalf("Expected error, got: %v", err)
+			}
+		})
+
+		t.Run("ReturnsCorrectlySerializedPayload", func(t *testing.T) {
+			massPayment := payload.NewMassPayment("user", "password", "signature", payload.ReceiverTypeEmail)
+			massPayment.Version = "1.0"
+			massPayment.EmailSubject = "Test email"
```
Comment (revett) [inline payload/mass_payment_test.go:34]: 🏄 Add new line under

**Review by revett — APPROVED** [review]
GTG after changes

**Pull request merged**